#Exercise 2

In [ ]:
!pip install -q pyomo

In [ ]:
from pyomo.environ import *

In [ ]:
import numpy as np

In [ ]:
coef = np.loadtxt('/content/lab4_ex2_lp_ip_coef.txt', delimiter = ',')

In [ ]:
print(coef.shape)
print('Num rows:',coef.shape[0])
print('Num cols:',coef.shape[1])

(6, 501)
Num rows: 6
Num cols: 501


In [ ]:
model = ConcreteModel()

In [ ]:
N = coef.shape[1]-1
M = coef.shape[0]-1

In [ ]:
obj_coef = coef[0][:-1]
print(obj_coef.shape)

(500,)


In [ ]:
constr_coef = coef[1:,:-1]
print(constr_coef.shape)

(5, 500)


In [ ]:
constr_rhs = coef[1:,-1]
print(constr_rhs.shape)

(5,)


In [ ]:
col_indices = np.arange(N)
model.x = Var(col_indices, domain = Binary)

In [ ]:
row_indices = np.arange(M)
model.constraints = ConstraintList()

In [ ]:
for i in row_indices:
  model.constraints.add(summation(constr_coef[i],model.x)<=constr_rhs[i])

In [ ]:
model.objective = Objective(expr = summation(obj_coef,model.x),sense = maximize)

In [ ]:
!apt-get install -y -qq coinor-cbc

In [ ]:
opt_cbc = SolverFactory('cbc')

In [ ]:
result_x = opt_cbc.solve(model)
print('Solver status:', result_x.solver.status)
print('Solver termination condition:',result_x.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
print('Objective = ', model.objective())

Objective =  295828.0


#Ans 3 :

The optimal value of the objective function is : 295828

In [ ]:
print('Solver time:',result_x.solver.time)

Solver time: 65.68236517906189


##Ans 4 :

The time taken to solve the milp is 65.68 seconds

In [ ]:
model_y = ConcreteModel()
model_y.y = Var(col_indices)
model_y.objective_y = Objective(expr = summation(obj_coef,model_y.y),sense = maximize)
model_y.constraints_y = ConstraintList()
for i in row_indices:
  model_y.constraints_y.add(summation(constr_coef[i],model_y.y)<=constr_rhs[i])
lower_bound = 0
upper_bound = 1
for j in col_indices:
  model_y.y[j].setlb(lower_bound)
  model_y.y[j].setub(upper_bound)

In [ ]:
result_y = opt_cbc.solve(model_y)
print('Solver status:', result_y.solver.status)
print('Solver termination condition:',result_y.solver.termination_condition)

Solver status: ok
Solver termination condition: optimal


In [ ]:
print('Objective = ', model_y.objective_y())

Objective =  295896.37697236


##Ans 5 :

The optimal value of the objective function in the case of LP is : 295896.377

In [ ]:
print('Solver time:',result_y.solver.time)

Solver time: 0.024901628494262695


##Ans 6 :

The time taken to solve the LP is 0.0249 seconds.

The time taken to solve the LP is way less than the time taken to solve the MILP.
295828, 295896
The optimal value for the LP is 295896.377 whereas for the MILP is 295828.



In [ ]:
print(result_x.solver.statistics.branch_and_bound)


Number of bounded subproblems: 347044
Number of created subproblems: 347044



In [ ]:
print(result_y.solver.statistics.branch_and_bound)


Number of bounded subproblems: None
Number of created subproblems: None



##Ans 7 :

The statistics printed in the branch and bound section of MILP is :

Number of bounded subproblems: 347044

Number of created subproblems: 347044

The statistics printed in the branch and bound section of LP is :

Number of bounded subproblems: None

Number of created subproblems: None